server/services/v1/key_generator.go

```diff
@@ -0,0 +1,131 @@
+}
+
+// get returns generated id for the supported primary key fields.
+func (g *generator) get(ctx context.Context, table []byte, field *schema.Field) ([]byte, error) {
+	switch field.Type() {
+	case schema.StringType, schema.UUIDType:
+		return []byte(fmt.Sprintf(`"%s"`, uuid.NewUUIDAsString())), nil
+	case schema.ByteType:
+		return jsoniter.Marshal([]byte(uuid.NewUUIDAsString()))
+	case schema.DateTimeType, schema.Int64Type, schema.Int32Type:
+		return g.generateInTx(ctx, table, field)
+	}
+	return nil, api.Errorf(codes.InvalidArgument, "unsupported type found in auto-generator")
+}
+
+// generateInTx is used to generate an id in a transaction. This is mainly used where there is no easy way to guarantee
+// uniqueness so what we are doing is reserving this id in storage before returning to the caller so that only one
+// id is assigned to one caller.
+func (g *generator) generateInTx(ctx context.Context, table []byte, field *schema.Field) ([]byte, error) {
+	for {
+		tx, err := g.txMgr.StartTx(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		var valueTS time.Time
+		var valueI64 int64
+		var valueI32 int32
+		switch field.Type() {
+		case schema.DateTimeType:
+			valueTS, _, err = g.generateTimestamp(ctx, tx, table)
+		case schema.Int64Type:
+			_, valueI64, err = g.generateTimestamp(ctx, tx, table)
+		case schema.Int32Type:
+			valueI32, err = g.generateInt(ctx, tx, table)
+		}
+		if err != nil {
+			_ = tx.Rollback(ctx)
+		}
+
+		if err = tx.Commit(ctx); err == nil {
+			if !valueTS.IsZero() {
+				return []byte(fmt.Sprintf(`"%s"`, valueTS.Format(time.RFC3339Nano))), nil
+			} else if valueI64 > 0 {
+				return []byte(fmt.Sprintf(`%d`, valueI64)), nil
+			} else if valueI32 > 0 {
+				return []byte(fmt.Sprintf(`%d`, valueI32)), nil
+			}
+		}
+		if err != kv.ErrConflictingTransaction {
+			return nil, err
+		}
+	}
+}
+
+// generateTimestamp stores values like this: "generator-<table>-timestamp_id-<tsNano>". The value is embedded in the
+// key to reduce the contention.
+// ToDo: cleanup old timestamps
+func (g *generator) generateTimestamp(ctx context.Context, tx transaction.Tx, table []byte) (time.Time, int64, error) {
+	tsTime := time.Now().UTC()
+	tsNano := time.Now().UTC().UnixNano()
+	key := keys.NewKey([]byte(generatorSubspaceKey), table, timestampIdKey, encoding.UInt32ToByte(uint32(tsNano)))
+
+	if err := tx.Insert(ctx, key, internal.NewTableData(encoding.UInt32ToByte(uint32(tsNano)))); err != nil {
+		return time.Time{}, 0, err
+	}
+
+	return tsTime, tsNano, nil
+}
+
+// generateInt as it is used to generate int32 value, we are simply maintaining a counter. There is a contention to
+// generate a counter if it is concurrently getting executed but the generation should be fast then it is best to start
+// with this approach.
+func (g *generator) generateInt(ctx context.Context, tx transaction.Tx, table []byte) (int32, error) {
+	key := keys.NewKey([]byte(generatorSubspaceKey), table, int32IdKey)
+	it, err := tx.Read(ctx, key)
```

Comment:
Looks like a point of contention.

Reply:
that is correct, it is a good starting point for int32. Later we can think about something else.